Revamp empty special-casing in LINQ

stephentoub opened the pull request:
Enumerable.Empty used to return Array.Empty. Towards the beginning of .NET Core, LINQ was imbued with an internal "partition" concept for flowing more information around between operators, and as part of that, Empty was changed to return a singleton instance of a specialized partition implementation. The upside of this was that methods typed to return IPartition could return the same singleton as Empty. There are multiple downsides, however. For one, the whole IPartition concept is only built into a "speed-optimized" build of LINQ; builds that care more about size (e.g. browser) end up not having it, and thus Empty there ends up being Array.Empty, such that a different type ends up being returned based on the build, which is not ideal. Further, any paths that check for empty now effectively have two things to check for: the empty partition or an empty array, making those checks more expensive, if they're even done at all, or in some cases missing out on possible optimization. This is more pronounced today, now that `[]` with collection expressions will produce Array.Empty, and it'd be really nice if there wasn't a difference between Enumerable.Empty and `[]` assigned to `IEnumerable<T>`.

This change puts Enumerable.Empty back to always being Array.Empty. The internal IPartition-based APIs that drove us to need the EmptyPartition are changed to just use null as an indication of empty. Places we were already checking for `is EmptyPartition` are changed to check for an empty array (if they weren't already), and other APIs that weren't checking at all now have a check if it makes sense to do so (I audited all of the APIs, and didn't include checks in ones where it could meaningfully affect semantics, e.g. a fast path that might cause us not to get an enumerator from a secondary enumerable input).

stephentoub committed:
Revamp empty special-casing in LINQ

Enumerable.Empty used to return Array.Empty. Towards the beginning of .NET Core, LINQ was imbued with an internal "partition" concept for flowing more information around between operators, and as part of that, Empty was changed to return a singleton instance of a specialized partition implementation. The upside of this was that methods typed to return IPartition could return the same singleton as Empty. There are multiple downsides, however. For one, the whole IPartition concept is only built into a "speed-optimized" build of LINQ; builds that care more about size (e.g. browser) end up not having it, and thus Empty there ends up being Array.Empty, such that a different type ends up being returned based on the build, which is not ideal. Further, any paths that check for empty now effectively have two things to check for: the empty partition or an empty array, making those checks more expensive, if they're even done at all, or in some cases missing out on possible optimization. This is more pronounced today, now that `[]` with collection expressions will produce Array.Empty, and it'd be really nice if there wasn't a difference between Enumerable.Empty and `[]` assigned to `IEnumerable<T>`.

This change puts Enumerable.Empty back to always being Array.Empty. The internal IPartition-based APIs that drove us to need the EmptyPartition are changed to just use null as an indication of empty. Places we were already checking for `is EmptyPartition` are changed to check for an empty array (if they weren't already), and other APIs that weren't checking at all now have a check if it makes sense to do so (I audited all of the APIs, and didn't include checks in ones where it could meaningfully affect semantics, e.g. a fast path that might cause us not to get an enumerator from a secondary enumerable input).

eiriktsarpalis reviewed (commented):
Seems to be causing test failures?

eiriktsarpalis commented on src/libraries/System.Linq/src/System/Linq/Enumerable.cs:
```diff
@@ -11,6 +11,19 @@ public static partial class Enumerable
+        /// It must not return true for things that are currently empty but could become non-empty, like a list.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsImmutableEmpty<TSource>(IEnumerable<TSource> source) =>
```
The method name and preceding comments seem to suggest a predicate that covers a broader range of types than what the implementation is doing. Shouldn't this just be `IsEmptyArray`?

stephentoub replied:
I wanted to accommodate other things in the future without a rename (this started out handling both empty arrays and the empty partition type). But I can rename it.

eiriktsarpalis replied:
Do we care enough about empty immutable arrays?

stephentoub replied:
> Do we care enough about empty immutable arrays?

Probably not.

eiriktsarpalis commented on src/libraries/System.Linq/src/System/Linq/Range.cs:
```diff
@@ -18,7 +18,7 @@ public static IEnumerable<int> Range(int start, int count)
             if (count == 0)
             {
-                return Empty<int>();
+                return [];
```
This would also make the returned type subject to compiler implementation. Should we perhaps be more prescriptive for IEnumerable types?

stephentoub replied:
> Should we perhaps be more prescriptive for IEnumerable types?

I don't want to guarantee the type that's returned. This doesn't seem different than accepting whatever type the compiler manufactures for an iterator. I think we're fighting an uphill battle if we discourage use of []. The only places I didn't use it were in places I couldn't (ie it wouldn't compile) or in Empty itself, as I can imagine the compiler might at some point return that.

stephentoub committed:
Merge branch 'main' into linqempty

stephentoub committed:
Rename IsImmutableEmpty to IsEmptyArray per PR feedback

stephentoub committed:
Remove bogus PLINQ tests

The tests were validating the underlying type of the operator returned for Concat, which is not material.

stephentoub committed:
Fix Skip special-casing

The check needs to be moved to before the count <= 0 check... otherwise calling Skip on an empty array with a count of 0 would still allocate an iterator.

stephentoub commented:
> Seems to be causing test failures?

Yup. Fixed.
